src/build.rs

```diff
@@ -263,3 +265,323 @@ fn build_nasm(root: &Path, vendor_dir: &Path, out_dir: &Path, target_arch: &str,
+        decls.push( Decl { name, is_pub, doc, attrs, args, ret_ty, lifetime });
+    }
+
+    decls
+}
+
+fn generate_cxx_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_cxx_wrapper(d)).collect()
+}
+
+fn generate_rust_wrappers(decls: &Vec<Decl>) -> Vec<String> {
+    decls.iter().map(|d| generate_rust_wrapper(d)).collect()
+}
+
+fn generate_rust_wrapper(decl: &Decl) -> String {
+    format!(r#"
+    {doc}
+    {attrs}
+    {pub_vis} unsafe fn {wrapper_name}{lifetime}({params}) -> {ret_ty} {{
+        extern "C" {{
+            fn {c_wrapper_name}{lifetime}({params}) -> WrapperResult<{ret_ty}>;
+        }}
+        let result = {c_wrapper_name}({arg_ids});
+        if result.discriminant == 0 {{
+           result.value.value
+        }} else {{
+           let WrapperError{{ buf, len, cap }}: WrapperError = result.value.error;
+           panic!(String::from_raw_parts(buf, len, cap))
+        }}
+    }}
+    "#,
+            doc = decl.doc.join("\n"),
+            attrs = decl.attrs.join("\n"),
+            pub_vis = if decl.is_pub { "pub" } else { "" },
+            wrapper_name = decl.name,
+            lifetime = decl.lifetime.clone().map(|l| format!("<{}>", l)).unwrap_or_default(),
+            params = decl.args.iter().map(|(id, ty)| format!("{}: {}", id, ty)).collect::<Vec<_>>().join(","),
+            c_wrapper_name = format!("{}_cxx", decl.name),
+            arg_ids = decl.args.iter().map(|(id, _)| id.clone()).collect::<Vec<_>>().join(","),
+            ret_ty = decl.ret_ty.clone().unwrap_or("()".to_string()),
+    )
+}
+
+fn generate_cxx_wrapper(decl: &Decl) -> String {
+    let ret_ty = decl.ret_ty.clone().map(|t| c_type(&t)).unwrap_or("void".to_string());
+    let arg_ids = decl.args.iter().map(|(id, _)| id.clone()).collect::<Vec<_>>().join(",");
+    let value_return = if decl.ret_ty.is_some() {
+        format!(
+            "return WrapperResult<{ret_ty}>::value({wrapper_name}({arg_ids}));",
+            ret_ty = ret_ty,
+            wrapper_name = decl.name,
+            arg_ids = arg_ids
+        )
+    } else {
+        format!(
+            "{wrapper_name}({arg_ids}); return WrapperResult<{ret_ty}>::value();",
+            ret_ty = ret_ty,
+            wrapper_name = decl.name,
+            arg_ids = arg_ids
+        )
+    };
+    format!(r#"
+    extern "C" WrapperResult<{ret_ty}> {c_wrapper_name}({params}) {{
+        try {{
+             {value_return}
+        }} catch(mozjpeg_rust_wrapper_exception const& e) {{
+             return WrapperResult<{ret_ty}>::error(e.error);
+        }}
+    }}
+    "#,
+            ret_ty = ret_ty,
+            c_wrapper_name = format!("{}_cxx", decl.name),
+            params = decl.args.iter().map(|(id, ty)| c_param(&id, &ty)).collect::<Vec<_>>().join(","),
+            value_return = value_return
+    )
+}
+
+fn is_c_ptr(v: &str) -> bool { v.starts_with("&") || v.starts_with("*") }
```

Comment:
I mostly adapted all this code to translate Rust types to C types from ctest, but that library does not expose these. You probably want to use a library for this.